src/Storages/MergeTree/ReplicatedMergeTreeAltersSequence.cpp

```diff
@@ -23,11 +23,13 @@ void ReplicatedMergeTreeAltersSequence::addMutationForAlter(int alter_version, s
 }
 
 void ReplicatedMergeTreeAltersSequence::addMetadataAlter(
-    int alter_version, bool have_mutation, std::lock_guard<std::mutex> & /*state_lock*/)
+    int alter_version, std::lock_guard<std::mutex> & /*state_lock*/)
 {
+    /// Data alter (mutation) always added before. See ReplicatedMergeTreeQueue::pullLogsToQueue.
+    /// So mutation alredy added to this sequence or doesn't exist.
     if (!queue_state.count(alter_version))
-        queue_state.emplace(alter_version, AlterState{.metadata_finished=false, .data_finished=!have_mutation});
-    else /// Data alter can be added before.
+        queue_state.emplace(alter_version, AlterState{.metadata_finished=false, .data_finished=true});
```

Comment:
Why may ALTER queries hung? Do they wait on `data_finished` being true?

Reply:
Yes, but corresponding data alter can be already killed.